### DISCUSSION TITLE
Issue when setting robot's initial position

### DISCUSSION THREAD

**Discussion by Xindictus** [gstavrinos/ai_msc_robotic_systems_assignments#44] [Q&A]
I have an issue when trying to set the robot's initial position.

I'm getting the robot's position from a topic and transforming the message to the required by the `nav2` docs `PoseStamped` message.

This fails with the following error:
```shell
Failed to transform initial pose in time (Lookup would require extrapolation into the future. 
Requested time 17.298000 but the latest data is at time 17.286000, when looking up transform
from frame [base_footprint] to frame [odom])
```

From a bit of a Google search, this seems like a kind of sync issue, but, it could always just be my code.

I have tried using both the navigator's and the node's "clock to message" function for the header's stamp, but, both return similar errors.

As a result the initial position is then set to something like:
```shell
Setting pose (17.298000): -0.000 0.000 -0.001
```

**Comment by gstavrinos** [comment]
First off, the robot's initial position should already be correct. In case it is not, you should, indeed set it.

Secondly, regarding your error, how are you trying to get the transform. It seems like you are requesting a "future" time. This is most probably due to requesting a `now()` time, for which the position of the robot is not yet known (by approximately 0.01 seconds according to the error)

Check the above and feel free to post some code :)

**Reply by Xindictus** [reply]
> First off, the robot's initial position should already be correct. In case it is not, you should, indeed set it.

That means we don't need to set it? Is this the same reason why `waitUntilNav2Active` fails with following error?
```shell
[ros2-39] [table_explorer.py-1] Exception in thread Thread-1 (run):
[ros2-39] [table_explorer.py-1] Traceback (most recent call last):
[ros2-39] [table_explorer.py-1]   File "/usr/lib/python3.10/threading.py", line 1016, in _bootstrap_inner
[ros2-39] [table_explorer.py-1]     self.run()
[ros2-39] [table_explorer.py-1]   File "/usr/lib/python3.10/threading.py", line 953, in run
[ros2-39] [table_explorer.py-1]     self._target(*self._args, **self._kwargs)
[ros2-39] [table_explorer.py-1]   File "/home/robotics/ros2_ws/install/tt_table_explorer/lib/tt_table_explorer/table_explorer.py", line 166, in run
[ros2-39] [table_explorer.py-1]     nav.waitUntilNav2Active()
[ros2-39] [table_explorer.py-1]   File "/opt/ros/humble/lib/python3.10/site-packages/nav2_simple_commander/robot_navigator.py", line 313, in waitUntilNav2Active
[ros2-39] [table_explorer.py-1]     self._waitForNodeToActivate(localizer)
[ros2-39] [table_explorer.py-1]   File "/opt/ros/humble/lib/python3.10/site-packages/nav2_simple_commander/robot_navigator.py", line 545, in _waitForNodeToActivate
[ros2-39] [table_explorer.py-1]     rclpy.spin_until_future_complete(self, future)
[ros2-39] [table_explorer.py-1]   File "/opt/ros/humble/local/lib/python3.10/dist-packages/rclpy/__init__.py", line 248, in spin_until_future_complete
[ros2-39] [table_explorer.py-1]     executor.spin_until_future_complete(future, timeout_sec)
[ros2-39] [table_explorer.py-1]   File "/opt/ros/humble/local/lib/python3.10/dist-packages/rclpy/executors.py", line 303, in spin_until_future_complete
[ros2-39] [table_explorer.py-1]     self.spin_once_until_future_complete(future, timeout_sec)
[ros2-39] [table_explorer.py-1]   File "/opt/ros/humble/local/lib/python3.10/dist-packages/rclpy/executors.py", line 746, in spin_once_until_future_complete
[ros2-39] [table_explorer.py-1]     self._spin_once_impl(timeout_sec)
[ros2-39] [table_explorer.py-1]   File "/opt/ros/humble/local/lib/python3.10/dist-packages/rclpy/executors.py", line 728, in _spin_once_impl
[ros2-39] [table_explorer.py-1]     handler, entity, node = self.wait_for_ready_callbacks(timeout_sec=timeout_sec)
[ros2-39] [table_explorer.py-1]   File "/opt/ros/humble/local/lib/python3.10/dist-packages/rclpy/executors.py", line 711, in wait_for_ready_callbacks
[ros2-39] [table_explorer.py-1]     return next(self._cb_iter)
[ros2-39] [table_explorer.py-1] ValueError: generator already executing
```

> Secondly, regarding your error, how are you trying to get the transform.

Currently, I'm getting the transform using the topic `/ground_truth_odom`. This returns an `Odometry` message which I'm then transforming to a `PoseStamped`.

> It seems like you are requesting a "future" time.

What does that mean? 🤣 

> This is most probably due to requesting a now() time, for which the position of the robot is not yet known (by approximately 0.01 seconds according to the error)

When I create the `PoseStamped` message, I just add the current stamp. Is that wrong? Should I use the stamp from the `Odometry` message? Although that doesn't makes sense to me.

**Reply by gstavrinos** [reply]
`waitUntilNav2Active` which is part of the `robot_navigator` script if I recall correctly is not part of the predefined solution of the exercise. Regardless, this error appears to be due to some kind of (recursive) mishandling of an object.

The `ground_truth_odom` topic, as its name suggests is a ground truth topic published by Gazebo and not navigation. Generally this would be considered "cheating" since you are not using the nav2 but an ideal topic. This is not a problem in our case, since you did not configure the navigation package, so you are not being tested on that.

The "future" time refers to the fact (exactly as the error suggests) that you are requesting a time that is not yet known the the transformations of the robot.

As a general not, when requesting a transformation, you should not request a `now()` time, but the latest available.


Regardless of all the above, I would point you towards using an action to call the nav2 goal and use its feedback for the progress of the operation.

**Reply by Xindictus** [reply]
I tried using `/amcl_pose` before `/ground_truth_odom`, but, I faced the issue mentioned in https://github.com/gstavrinos/ai_msc_robotic_systems_assignments/discussions/43.

> since you are not using the nav2 but an ideal topic

I will check out what you mean by that and come back.

> when requesting a transformation, you should not request a now() time, but the latest available.

Which means I should be using the stamp I got from the whatever topic?

> Regardless of all the above, I would point you towards using an action to call the nav2 goal and use its feedback for the progress of the operation.

I have already tested that with random values and it's clear. I just wanted to resolve the above issues first.

**Reply by gstavrinos** [reply]
You should be using the latest available transformation time. The stamp on the topic is not necessarily the same. Check the documentation of the function you use to get your transformation. It should have a "latest available time" option.

Regarding the solution in general, consider using a transform listener instead of topics. It might make your life easier!

**Reply by Xindictus** [reply]
I probably don't understand what I need to do. I have the following (changed from topic to getting the transform from TF tree):
```python
self.request = Empty.Request()
nav = BasicNavigator()
initial_pose = PoseStamped()

transform = self.tf_buffer.lookup_transform(
    'map',
    'odom',
    rclpy.time.Time(seconds=0)
)
# Copy the header
initial_pose.header = transform.header

# Copy the translation
initial_pose.pose.position.x = transform.transform.translation.x
initial_pose.pose.position.y = transform.transform.translation.y
initial_pose.pose.position.z = transform.transform.translation.z

# Copy the rotation
initial_pose.pose.orientation = transform.transform.rotation

nav.setInitialPose(initial_pose)
```

I still get the same error:
```shell
Failed to transform initial pose in time (Lookup would require extrapolation into the future.  Requested time 16.865000 but the latest data is at time 16.862000, when looking up transform from frame [base_footprint] to frame [odom])
```

I have tried other frame ids as well, but, error is similar.

**Reply by gstavrinos** [reply]
The way you approach this with the `tf_buffer` seems correct. The frames are not though. Both `map` and `odom` are reference frames that are "static" and do not move, but rather used to reference other moving parts of the system.

Regardless, you shouldn't be getting this error during lookup. Check if running the simulation headless helps. Also, check if anything fails to execute or crashing during initialization. I will create a dummy node with a tf listener and look into it when I get home.

**Reply by Xindictus** [reply]
The actual frames I'm using are `map` and `base_footprint`. These were just an example. I've taken the frame id by exporting the TF tree into pdf during the sim runtime.

Wrt the error, this is not coming from the lookup. It's coming from `nav.setInitialPose(initial_pose)`. The lookup works fine and I already had it working for some other frame ids using a timer to retry until they are available.

My main issue is the initial pose. That is where I'm getting the error. Is the following by any chance setting the initial pose?
```python
self.request = Empty.Request()
```

**Comment by gstavrinos** [comment]
Have tried assigning a goal to the robot without setting its initial pose? (It  should appear to be correctly localized in rviz anyway)

The `Empty` request should be used to notify the umpire when you reach each waypoint. It normally does nothing apart from initializing the `Empty` service request.

**Reply by chaidop** [reply]
May I also ask just to be clear (I am having same issue understanding the localization part). So tounderstand clearly, the localization part is kind of done already by the scripts. This means, we do not need to set an inital_pose or listen to the pose of the robot through let's say amcl. We only need to give a goal Pose and it will be calculated on its own? 
Sorry if it is dumb question I think I did not catch it correctly

**Comment by gstavrinos** [comment]
Setting the initial pose of the robot is required when the robot's pose is not correct. By checking the robot's pose using rviz, you should be able to see that is correctly initialized. You can easily evaluate the robot's pose by checking that the laserscans align with the walls of the map. If the robot is not correctly localized, you should fix its pose. If it's correctly localized, you can proceed with navigation. Just give it a go! :robot:
